### ISSUE TITLE
1.2: Can't use ES2015 code inside of package.js

### ISSUE THREAD

**Issue by mizzao** [meteor/meteor#5240]
e.g. in https://github.com/mizzao/meteor-openlayers/, the following code in `Package.onUse`:

```js
Package.onUse(function (api) {
  const path = Npm.require('path');

  const files = [
    "blank.gif",
    "cloud-popup-relative.png",
    "drag-rectangle-off.png",
    "drag-rectangle-on.png",
    "east-mini.png",
    "layer-switcher-maximize.png",
    "layer-switcher-minimize.png",
    "marker-blue.png",
    "marker-gold.png",
    "marker-green.png",
    "marker.png",
    "measuring-stick-off.png",
    "measuring-stick-on.png",
    "north-mini.png",
    "panning-hand-off.png",
    "panning-hand-on.png",
    "slider.png",
    "south-mini.png",
    "west-mini.png",
    "zoombar.png",
    "zoom-minus-mini.png",
    "zoom-plus-mini.png",
    "zoom-world-mini.png"
  ];

  for (let file of files) {
    api.addAssets(path.join("openlayers", "img", file), 'client');
  }

});
```

throws the following error:

```
While reading package from `/home/mao/projects/CrisisMapping/CrowdMapper/packages/mizzao:openlayers`:
vm.js:32:12: Unexpected identifier
at new Script (vm.js:32:12)
at Function.Script.createScript (vm.js:48:10)
at Object.files.runJavaScript (/tools/fs/files.js:933:32)
at [object Object]._.extend.initFromPackageDir (/tools/isobuild/package-source.js:975:13)
at /tools/packaging/catalog/catalog-local.js:286:23
at /tools/utils/buildmessage.js:356:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:349:34
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:347:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at Object.enterJob (/tools/utils/buildmessage.js:321:26)
at initSourceFromDir (/tools/packaging/catalog/catalog-local.js:273:20)
at /tools/packaging/catalog/catalog-local.js:345:9
at /tools/utils/buildmessage.js:586:26
at Array.forEach (native)
at Function._.each._.forEach
(/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
at /tools/utils/buildmessage.js:584:9
at /tools/utils/buildmessage.js:356:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:349:34
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:347:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at enterJob (/tools/utils/buildmessage.js:321:26)
at Object.forkJoin (/tools/utils/buildmessage.js:546:3)
at [object Object]._.extend._loadLocalPackages (/tools/packaging/catalog/catalog-local.js:341:18)
at [object Object]._.extend.initialize (/tools/packaging/catalog/catalog-local.js:86:10)
at /tools/project-context.js:573:29
at /tools/utils/buildmessage.js:356:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:349:34
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:347:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at Object.enterJob (/tools/utils/buildmessage.js:321:26)
at /tools/project-context.js:565:20
at /tools/packaging/catalog/catalog.js:100:5
at /tools/utils/buildmessage.js:268:13
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:261:29
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:259:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:250:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at Object.capture (/tools/utils/buildmessage.js:249:19)
at Object.catalog.runAndRetryWithRefreshIfHelpful (/tools/packaging/catalog/catalog.js:99:31)
at ProjectContext._.extend._initializeCatalog (/tools/project-context.js:564:13)
at /tools/project-context.js:257:35
at /tools/utils/buildmessage.js:356:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:349:34
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:347:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at Object.enterJob (/tools/utils/buildmessage.js:321:26)
at ProjectContext._.extend._completeStagesThrough (/tools/project-context.js:247:18)
at ProjectContext._.extend.resolveConstraints (/tools/project-context.js:226:10)
at /tools/cli/commands-packages.js:1644:22
at /tools/utils/buildmessage.js:268:13
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:261:29
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:259:18
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at /tools/utils/buildmessage.js:250:23
at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
at Object.capture (/tools/utils/buildmessage.js:249:19)
at Object.main.captureAndExit (/tools/cli/main.js:268:29)
at Command.main.registerCommand.name (/tools/cli/commands-packages.js:1642:8)
at /tools/cli/main.js:1378:23
```

The error goes away if the `for...of` loop is removed. This is mysterious as I've stared at it a bit and don't actually see anything wrong with the code, so I'm not sure why it is crashing.

**Comment by mizzao** [comment]
According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of this should be a valid statement in ES2015.

**Comment by mizzao** [comment]
Here's another similar error I got from trying to update https://github.com/mizzao/meteor-build-fetcher:

```
   While loading plugin `fetcher` from package `mizzao:build-fetcher`:
   vm.js:32:12: Unexpected reserved word
   at new Script (vm.js:32:12)
   at Function.Script.createScript (vm.js:48:10)
   at Object.files.runJavaScript (/tools/fs/files.js:933:32)
   at /tools/isobuild/bundler.js:1325:15
   at Array.forEach (native)
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
   at JsImage.load (/tools/isobuild/bundler.js:1256:7)
   at /tools/isobuild/isopack.js:503:16
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at /tools/isobuild/isopack.js:493:20
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
   at [object Object]._.extend.ensurePluginsInitialized (/tools/isobuild/isopack.js:481:7)
   at /tools/isobuild/compiler.js:365:16
   at Array.forEach (native)
   at /tools/isobuild/compiler.js:364:26
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at compileUnibuild (/tools/isobuild/compiler.js:360:16)
   at /tools/isobuild/compiler.js:175:26
   at Array.forEach (native)
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
   at Object.compile (/tools/isobuild/compiler.js:171:5)
   at /tools/isobuild/isopack-cache.js:271:30
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at [object Object]._.extend._loadLocalPackage (/tools/isobuild/isopack-cache.js:227:18)
   at /tools/isobuild/isopack-cache.js:166:14
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at [object Object]._.extend._ensurePackageLoaded (/tools/isobuild/isopack-cache.js:158:20)
   at ensureLoaded (/tools/isobuild/isopack-cache.js:133:12)
   at /tools/isobuild/isopack-cache.js:160:11
   at Array.forEach (native)
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
   at /tools/isobuild/isopack-cache.js:159:11
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at [object Object]._.extend._ensurePackageLoaded (/tools/isobuild/isopack-cache.js:158:20)
   at ensureLoaded (/tools/isobuild/isopack-cache.js:133:12)
   at /tools/isobuild/isopack-cache.js:160:11
   at Array.forEach (native)
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:79:11)
   at /tools/isobuild/isopack-cache.js:159:11
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at [object Object]._.extend._ensurePackageLoaded (/tools/isobuild/isopack-cache.js:158:20)
   at /tools/isobuild/isopack-cache.js:73:14
   at /tools/packaging/package-map.js:57:7
   at Function._.each._.forEach
   (/home/mao/.meteor/packages/meteor-tool/.1.1.8.1bmxbnb++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/underscore/underscore.js:87:22)
   at [object Object]._.extend.eachPackage (/tools/packaging/package-map.js:49:7)
   at [object Object]._.extend.buildLocalPackages (/tools/isobuild/isopack-cache.js:72:24)
   at /tools/project-context.js:735:25
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at ProjectContext._.extend._buildLocalPackages (/tools/project-context.js:734:18)
   at /tools/project-context.js:257:35
   at /tools/utils/buildmessage.js:356:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:349:34
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:347:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.enterJob (/tools/utils/buildmessage.js:321:26)
   at ProjectContext._.extend._completeStagesThrough (/tools/project-context.js:247:18)
   at ProjectContext._.extend.saveChangedMetadata (/tools/project-context.js:235:10)
   at ProjectContext._.extend.prepareProjectForBuild (/tools/project-context.js:240:10)
   at /tools/runners/run-app.js:517:29
   at /tools/utils/buildmessage.js:268:13
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:261:29
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:259:18
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at /tools/utils/buildmessage.js:250:23
   at [object Object]._.extend.withValue (/tools/utils/fiber-helpers.js:114:14)
   at Object.capture (/tools/utils/buildmessage.js:249:19)
   at bundleApp (/tools/runners/run-app.js:516:31)
   at [object Object]._.extend._runOnce (/tools/runners/run-app.js:594:35)
   at [object Object]._.extend._fiber (/tools/runners/run-app.js:858:28)
   at /tools/runners/run-app.js:396:12
```

Unclear where this error is coming from, but it just seems to dislike random parts of Javascript code which are very hard to catch.

**Comment by mizzao** [comment]
Traceur seems to be absurdly buggy compared to Babel...

**Comment by benjamn** [comment]
> Apparently it's impossible to use ES6 code in package.js?

That's right, because we don't even know if a package `api.use`s the `ecmascript` package until we evaluate its `package.js` file, and the `ecmascript` package itself is implemented as a package.

The good news is that `package.js` will always be evaluated by the version of Node that we ship on the server, so some ES2015+ features will become natively available as we upgrade Node (see https://github.com/meteor/meteor/issues/5124).

**Comment by abernix** [comment]
Closing as the situation here was explained.  Node 4.4.7 is now included in Meteor 1.4 and you should find increased support for things like this as they are supported by the engine.  For example, the original issue should be resolved now as `for..of` is supported by Node 4.4.7.: http://node.green/#for--of-loops
